File: app/src/main/java/co/k2/newsbits/data/source/NewsRepository.java
package co.k2.newsbits.data.source;

import java.util.List;

import javax.inject.Inject;

import co.k2.newsbits.data.models.Article;
import co.k2.newsbits.data.source.local.LocalDataSourceImpl;
import co.k2.newsbits.data.source.remote.RemoteDataSource;
import io.reactivex.Completable;
import io.reactivex.Single;

/**
 * Copyright (C) 2019 K2 CODEWORKS
 * All rights reserved
 *
 * @author Kislay
 * @since 28/03/19
 */

public class NewsRepository {

    private final RemoteDataSource remoteDataSource;
    private final LocalDataSourceImpl localDataSource;

    @Inject
    public NewsRepository(RemoteDataSource remoteDataSource, LocalDataSourceImpl localDataSourceImpl) {
        this.remoteDataSource = remoteDataSource;
        this.localDataSource = localDataSourceImpl;
    }

    public Single<List<Article>> getCachedNews() {
        return localDataSource.getHeadlines(null);
    }

    public Single<List<Article>> fetchNews(String country) {
        return remoteDataSource.getHeadlines(country);
    }

    public Completable saveToCache(List<Article> articles) {
        return localDataSource.saveToCache(articles);
    }

    public Completable clearCache() {
        return localDataSource.clearCache();
    }

}

File: app/src/main/java/co/k2/newsbits/data/source/NewsRepository.kt
package co.k2.newsbits.data.source

import co.k2.newsbits.common.ApiResult
import co.k2.newsbits.common.ErrorObj
import co.k2.newsbits.data.models.ApiResponse
import co.k2.newsbits.data.models.Article
import co.k2.newsbits.data.models.ArticleCacheMetaData
import co.k2.newsbits.data.source.local.LocalDataSourceImpl
import co.k2.newsbits.data.source.remote.RemoteDataSourceImpl
import javax.inject.Inject

/**
 * Copyright (C) 2019 K2 CODEWORKS
 * All rights reserved
 *
 * @author Kislay
 * @since 28/03/19
 */
class NewsRepository @Inject constructor(
    private val remoteDataSource: RemoteDataSourceImpl,
    private val localDataSource: LocalDataSourceImpl
) {

    suspend fun cachedNews(): List<Article> {
        return localDataSource.getHeadlines(null)
    }

    suspend fun fetchNews(country: String): ApiResult<ApiResponse, ErrorObj> {
        return remoteDataSource.getHeadlines(country)
    }

    suspend fun saveToCache(articles: List<Article>, time: Long, country: String) {
        localDataSource.saveToCache(articles)
        localDataSource.saveCacheMetaData(ArticleCacheMetaData(time, country))
    }

    suspend fun clearCache() {
        return localDataSource.clearCache()
    }

    suspend fun getCacheMetadata(): ArticleCacheMetaData? {
        return localDataSource.getCacheMetaData()
    }

}
